Sabrina/exported function signature

sabrina-li opened the pull request:
Sorry for the large PR, most of the changes are due to auto generation of the API interfaces

I would suggest starting from `README` and `src/users.ts` + `src/events.ts` to see the changes in how clients may change the requests to provide per-request options (such as idempotency key or integration_source).

The goal is to remove any pre-request `options` from the functions signature and makes it so that we can add any parameters in the server APIs and will not cause a non-backward compatible change. Example:
```
// i.e. current signature
users.update(id:string, body:Request, include_schema: bool, opts: FSRequestOptions)
// if a new query parameter is added, the method becomes
users.update(id:string, body:Request, include_schema: bool, new_field: string, opts: FSRequestOptions)
// notice the parameters changed in a way that can break if `opts` is provided

// changes in this PR makes the functions
users.update(request: { id:string, body:Request, include_schema: bool })
// if a new query parameter is added, the method becomes
users.update(request: {id:string, body:Request, include_schema: bool, new_field:"won't break"})
```

Additionally a `withOptions` method is added that can override options if needed

mattgrosvenor-fs commented on README.md:
````diff
@@ -126,6 +128,15 @@ const fsClient = init({ apiKey: '<YOUR_API_KEY>' });
+
+  ```ts
+  const createResponse = await users
+      .withRequestOptions({ idempotencyKey: 'YOUR_KEY' })
````
I like being able to set options like this, very clean.

sabrina-li replied:
Yeah I kinda went back and forth with weather to introduce `withRequestOptions` or just have the alternate option where the request object like `users.update(request: {...}, { idempotencyKey: 'YOUR_KEY' })`..

sabrina-li replied:
I guess we could do both: users.withRequestOptions( { idempotencyKey: 'YOUR_KEY' } ).update({id:string, body:Request, include_schema: bool})..

mattgrosvenor-fs commented:
It seems like you lost the end of your description.

I favor the alternate option
```
users.update(request: {id:string, body:Request, include_schema: bool}, opts: FSRequestOptions)
```
I have used a style like this in the past, it makes it very easy to pass parameters without having to worry about order and modify properties being passed.

sabrina-li commented on src/api/events/EventsApi.ts:
```diff
@@ -35,7 +35,8 @@ export class EventsApi {
      * @param body
     */
-    public async createEvents(body: CreateEventsRequest, options?: FSRequestOptions): Promise<FSResponse<void>> {
+    public async createEvents(request: { body: CreateEventsRequest,  }): Promise<FSResponse<void>> {
```
Maybe overthinking, but this could be changed from `body` to `payload` with a bit more changes in the template.

sabrina-li commented on src/events.ts:
```diff
@@ -24,8 +24,11 @@ export interface EventsApi {
+            body: CreateBatchEventsImportJobRequest,
+            includeSchema?: boolean,
+        },
+        jobOptions?: BatchJobOptions,
```
I left `jobOptions` out of the `withOptions`, since it's lives on the job object and isn't really affect the "requests"

sabrina-li commented:
So I dropped `options` completely from the api functions signature and added `withOptions`
`withOptions` returns the `Events` or `Users` API, so it can't be chained more than once

Usage example:
```
events.withOptions(...).update({
  id: "123",
  body: {...reques}
});
```
or 
```
users.get({
  id: "123",
  includeSchema: true 
});
```

LMKWDYT

mattgrosvenor-fs commented:
> LMKWDYT

I don't think there is a need to chain `withOptions`, so that sounds good to me.

I like either setting up the options initially or overriding them on a per request basis as you have it.

mattgrosvenor-fs reviewed (approved):
Left a minor suggestion, LGTM
